src/codegen/x86_64/arch_x86_64.c

```diff
@@ -1242,11 +1219,11 @@ static void emit_instruction(CodegenContext *context, IRInstruction *inst) {
     else {
       enum RegSize size = -1;
       // TODO: Should this array to pointer decay happen here? Or higher up in codegen?
-      // TODO: type_sizeof(t_pointer) or something to load a pointer sized thing.
-      //       WE SHOULD NOT USE t_integer here!!
-      if (inst->operand->type->kind == TYPE_ARRAY) size = regsize_from_bytes(type_sizeof(t_integer));
-      // TODO: Is this right? Do we need to get size of pointed to type?
+      if (inst->operand->type->kind == TYPE_ARRAY) size = regsize_from_bytes(type_sizeof(t_pointer));
+      // TODO: We are "supposed" to be loading sizeof pointed to type
+      // here, but that causes segfaults when handling arrays.
```

Comment:
Probably for arrays, you have to do `array.of` rather than `pointer.to`. Is that the problem here? I don’t really want to think about this array/pointer stuff here rn because of what I mentioned earlier in an extensive comment regarding array loading (in `codegen.c` I believe).

Reply:
And I feel like if we can resolve that somehow then this should ideally just fall into place

Reply:
This has to do with only having one way that an array load is emitted, but two ways that it needs to be loaded

```
foo : integer[2]
ptr : @integer = foo[0] ;; or &foo ... needs to emit load of pointer to first element of `foo`
bar : integer[2]
bar := foo ;; needs to emit load of entire `foo`
```

Either way, the generated variable access of a subscript (IR_LOAD) needs to not actually load the array, but load a pointer to the array's first element, and then add the calculated subscript amount to that. You can see how the subscript loads `foo` first, but that emits a load of the entire array. So this is why we have hacks like this right now. This is also because `IR_LOAD` strips one level of indirection, which means the static_ref or alloca being a `@integer[2]` doesn't help, either

Reply:
Yeah, I commented on that problem in codegen.c